src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4069,8 +4076,21 @@ internal uint GetValEscape(BoundExpression expr, uint scopeOfTheContainingExpres
                             isRefEscape: false);
                     }
 
-                    return Math.Max(GetValEscape(binary.Left, scopeOfTheContainingExpression),
-                                    GetValEscape(binary.Right, scopeOfTheContainingExpression));
+#if DEBUG
+                    // Nested binary operators on the left side are not visited by BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator.
```

Comment:
>Nested binary operators on the left side are not visited by BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator. [](http://example.com/codeflow?start=23&length=131)

It is not obvious to me that not visiting a user-defined binary operator on the left is the right thing to do for the purpose of this visitor. The fact that the base class skips it completely doesn't mean that doing the same is appropriate for every derived visitor.

Reply:
Note, BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator also does similar trick for calls. Should they be treated specially too?

Reply:
> Note, BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator also does similar trick for calls. Should they be treated specially too?

That does not seem needed, as the ref safety analysis doesn't seem to (even indirectly) call AssertVisited on the receiver like it does for binary operator (where escape of the left-hand side is checked which in turn calls AssertVisited on it)

Reply:
However, it is true we need special handling in other methods than just GetValEscape, for example CheckValEscape

Reply:
> It is not obvious to me that not visiting a user-defined binary operator on the left is the right thing to do for the purpose of this visitor. The fact that the base class skips it completely doesn't mean that doing the same is appropriate for every derived visitor.

RefSafetyAnalysis does not currently override VisitBinaryOperator, so assuming that's correct, the optimization seems valid, i.e., it should be transparent to RefSafetyAnalysis that only the operands are visited, not the operators themselves.

Note that the handling of binary operators (and also the special handling of user-defined binary operators) doesn't happen in VisitBinaryOperator, it happens in GetValEscape/CheckValEscape, so it's unrelated to the walker's visiting behavior.